### PR TITLE
Bugfixes and tweaks for objectives and gamemode 

### PR DESCRIPTION
**Pull request by caelaislinn**
See commit log. Main changes are to disable spawning in for jobs without safe atmos and adding a slipspace engine for the covvie ship.

### CODE REVIEW

File: code/modules/client/preferences_spawnpoints.dm [hunk 1]
```diff
@@ -19,6 +19,8 @@ GLOBAL_VAR(spawntypes)
 	var/list/restrict_job_type = null
 	var/list/disallow_job = null
 	var/list/disallow_job_type = null
+	var/disable_atmos_unsafe = 1
```
Comment (Kmc2000) [inline code/modules/client/preferences_spawnpoints.dm:22]: Why not TRUE

File: code/modules/client/preferences_spawnpoints.dm [hunk 2]
```diff
@@ -35,7 +37,26 @@ GLOBAL_VAR(spawntypes)
...
-	if(!turfs || !turfs.len)
+	if(!turfs)
+		return 0
+
+	if(disable_atmos_unsafe)
+		var/list/newly_dangerous_turfs = list()
+		if(!unsafe_turfs)
+			unsafe_turfs = list()
+		if(turfs)
+			for(var/turf/T in turfs)
```
Comment (Kmc2000) [inline code/modules/client/preferences_spawnpoints.dm:48]: Would it not be easier to make one turfs list and loop thru it there?
Also, typeless loop would be more efficient

File: code/modules/halo/covenant/slipspace.dm [hunk 3]
```diff
@@ -8,3 +8,71 @@
 	density = 1
 	bound_width = 64
 	bound_height = 64
+	var/slipspace_activate_timer = 10 SECONDS
+	var/slipspace_chargeup_timer = 60 SECONDS
+	var/slipspace_jump_time = 0
+
+/obj/structure/covenant_slipspace/allowed(var/mob/user)
+	var/mob/living/carbon/human/h = user
```
Comment (Kmc2000) [inline code/modules/halo/covenant/slipspace.dm:16]: You may want a if(!H) early return

File: code/modules/halo/covenant/slipspace.dm [hunk 4]
```diff
@@ -8,3 +8,71 @@
...
+						slipspace_jump_time = world.time + slipspace_chargeup_timer
+						GLOB.processing_objects += src
+						log_admin("[user] the [user.mind.assigned_role] (CKEY: [user.ckey]) activated the slipspace engine. Jump timer: [slipspace_chargeup_timer / 10] seconds.")
+						if(istype(ship))
+							to_chat(user,"<span class='notice'>[src] has been activated. [ship] will enter slipspace in [slipspace_chargeup_timer / 10] seconds.</span>")
+	else
+		to_chat(user,"<span class='warning'>You are unable to decipher how [src] works.</span>")
+
+/obj/structure/covenant_slipspace/process()
+	if(slipspace_jump_time && world.time > slipspace_jump_time)
```
Comment (Kmc2000) [inline code/modules/halo/covenant/slipspace.dm:47]: (  >=?)

File: code/modules/halo/covenant/slipspace.dm [hunk 5]
```diff
@@ -8,3 +8,71 @@
...
+
+/obj/effect/slipspace_rupture
+	name = "slipspace rupture"
+	icon = 'code/modules/halo/covenant/slipspace.dmi'
+	icon_state = "slipspace_effect"
+	pixel_x = -16
+	pixel_y = -16
+
+/obj/effect/slipspace_rupture/New()
+	spawn(6)
```
Comment (Kmc2000) [inline code/modules/halo/covenant/slipspace.dm]: Bad

File: code/modules/halo/covenant/slipspace.dm [hunk 6]
```diff
@@ -8,3 +8,71 @@
...
+				var/headingdir = ship.get_heading()
+				if(!headingdir)
+					headingdir = ship.dir
+				var/turf/T = ship.loc
+				for(var/i=0, i<6, i++)
+					T = get_step(T,headingdir)
+				new /obj/effect/slipspace_rupture(T)
+				for(var/i=0, i<6, i++)
+					ship.loc = get_step(ship, headingdir)
+					sleep(1)
```
Comment (Kmc2000) [inline code/modules/halo/covenant/slipspace.dm]: Can you not just CHECK_TICK?
Sleeps are less than ideal

**Review by BDpuffy420 — APPROVED** [review]

File: code/modules/client/preferences_spawnpoints.dm (hunk 2 above)
Comment (caelaislinn) [reply code/modules/client/preferences_spawnpoints.dm:48]: This updates the list here. It's necessary to do 2 passes

File: code/modules/client/preferences_spawnpoints.dm (hunk 1 above)
Comment (caelaislinn) [reply code/modules/client/preferences_spawnpoints.dm:22]: I have literally never used TRUE/FALSE macros in byond

File: code/modules/halo/covenant/slipspace.dm (hunk 3 above)
Comment (caelaislinn) [reply code/modules/halo/covenant/slipspace.dm:16]: istype() will do a null check, but there is a potential null access here so I put up a fix

File: code/modules/halo/covenant/slipspace.dm (hunk 4 above)
Comment (caelaislinn) [reply code/modules/halo/covenant/slipspace.dm:47]: Doesn't matter, the process tick will almost never align with the decisecond target time and if they do it's only a couple of seconds delay

**Comment by caelaislinn** [comment]
Partial #1070

**Pull request merged**